cmd/publicize/server.go

```diff
@@ -0,0 +1,154 @@
+	}
+	pluginHelp.AddCommand(pluginhelp.Command{
+		Usage:       "/publicize",
+		Description: "Merge the specific pull request into the bonded repository",
+		WhoCanUse:   "Members of the trusted organization for the repo.",
+		Examples:    []string{"/publicize"},
+	})
+	return pluginHelp, nil
+}
+
+type server struct {
+	githubTokenGenerator func() []byte
+
+	gitName  string
+	gitEmail string
+
+	bonds map[string]string
+
+	ghc github.Client
+	gc  git.ClientFactory
+}
+
+func (s *server) handleIssueComment(l *logrus.Entry, ic github.IssueCommentEvent) {
+	if !publicizeRe.MatchString(ic.Comment.Body) {
+		return
+	}
+
+	org := ic.Repo.Owner.Login
+	repo := ic.Repo.Name
+	num := ic.Issue.Number
+	commentAuthor := ic.Comment.User.Login
+
+	logger := logrus.WithFields(logrus.Fields{
+		github.OrgLogField:  org,
+		github.RepoLogField: repo,
+		github.PrLogField:   num,
+	})
+
+	if !ic.Issue.IsPullRequest() {
+		s.createComment(ic, fmt.Sprintf("Publicize plugin can only be used in pull requests"), logger)
+		return
+	}
+
+	// Only org members should be able to publicize a pull request.
+	ok, err := s.ghc.IsMember(org, commentAuthor)
+	if err != nil {
+		return
+	}
+	if !ok {
+		s.createComment(ic, fmt.Sprintf("only [%s](https://github.com/orgs/%s/people) org members may request a publicize", org, org), logger)
+		return
+	}
+	logger.Info("Publicize of PR has been requested.")
+
+	pr, err := s.ghc.GetPullRequest(org, repo, num)
+	if err != nil {
+		return
+	}
+	baseBranch := pr.Base.Ref
+
+	if !pr.Merged {
+		s.createComment(ic, "cannot publicize an unmerged PR", logger)
+		return
+	}
+
+	logger.Info("Searching bond...")
+	bondedOrgRepo, ok := s.bonds[fmt.Sprintf("%s/%s", org, repo)]
+	if !ok {
+		logger.Warn("There is no public repository configured for the current repository.")
+		s.createComment(ic, "cannot publicize because there is no bond for the current pull request", logger)
+		return
+	}
+
+	bondedOrg := strings.Split(bondedOrgRepo, "/")[0]
+	bondedRepo := strings.Split(bondedOrgRepo, "/")[1]
+
+	logger.Infof("Trying to merge the PR to destination: %s/%s@%s", bondedOrg, bondedRepo, baseBranch)
+	if err := s.mergeAndPushToRemote(bondedOrg, bondedRepo, org, repo, baseBranch); err != nil {
+		logger.WithError(err).Warnf("couldn't merge the pull request and push to the destination: %v", err)
+		s.createComment(ic, fmt.Sprintf("Publicize failed with error: %v", err), logger)
+		return
+	}
+
+	bondedOrgRepoLink := fmt.Sprintf("https://github.com/%s", bondedOrgRepo)
+	s.createComment(ic, fmt.Sprintf("Publicize to the repository: [%s/%s](%s) was successful",
+		bondedOrg, bondedRepo, bondedOrgRepoLink), logger)
+}
+
+func (s *server) mergeAndPushToRemote(bondedOrg, bondedRepo string, org, repo, branch string) error {
+	defer func() {
+		if err := s.gc.Clean(); err != nil {
+			logrus.WithError(err).Error("Could not clean up git client cache.")
+		}
+	}()
+
+	repoClient, err := s.gc.ClientFor(bondedOrg, bondedRepo)
```

Comment:
What if we are publicizing multiple PRs for a single repo? Won't they race over a single shared git clone?

Reply:
Note that this is probably not that uncommon - I can imagine a secuity fix being fixed in multiple PRs. For large repos the publicize can take some time (because for merge you need full fetch), and I can imagine the responsible person will just go to all PRs and `/publicize` in all of them.

Reply:
I don't think that this will race. I think the multiple goroutines will clone the same repo but in the different tmp folder. I will dig this to be sure.

Reply:
It seems that the repos will be cloned in different folder. ref: https://github.com/kubernetes/test-infra/blob/master/prow/git/v2/client_factory.go#L245